packages/arb-avm-cpp/data_storage/src/datastorage.cpp

```diff
@@ -220,3 +260,46 @@ rocksdb::Status DataStorage::updateSecretHashSeed() {
     }
     return status;
 }
+
+ConcurrentCounter DataStorage::getCounter() const {
+    return ConcurrentCounter::Get(this);
+}
+
+ConcurrentCounter::ConcurrentCounter(const DataStorage* _storage)
+    : storage(_storage) {
+    storage->concurrent_database_access_counter++;
+    if (storage->shutting_down) {
+        // Destructor will take care of decrementing counter
+        throw DataStorage::shutting_down_exception();
+    }
+}
+
+ConcurrentCounter::~ConcurrentCounter() {
+    if (storage == nullptr) {
+        return;
+    }
+    auto new_counter = storage->concurrent_database_access_counter.fetch_sub(1);
+    assert(new_counter >= 0);
+    (void)new_counter;  // silence unused variable warning on release builds
+                        // where assert is disabled
+}
+
+ConcurrentCounter ConcurrentCounter::Get(const DataStorage* storage) {
+    if (storage->shutting_down) {
+        throw DataStorage::shutting_down_exception();
+    }
+    return ConcurrentCounter(storage);
+}
+
+ConcurrentCounter::ConcurrentCounter(ConcurrentCounter&& other) noexcept
+    : storage(other.storage) {
+    other.storage = nullptr;
+}
+
+ConcurrentCounter& ConcurrentCounter::operator=(
```

Comment:
Maybe even simpler to just delete this method? Seems like we probably don't need assignment

Reply:
Doesn't hurt anything, and good to have implementation to make sure nothing bad happens in the future.